client/object_search.go

```diff
@@ -0,0 +1,266 @@
+package client
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"errors"
+	"fmt"
+	"io"
+
+	v2object "github.com/nspcc-dev/neofs-api-go/v2/object"
+	v2refs "github.com/nspcc-dev/neofs-api-go/v2/refs"
+	rpcapi "github.com/nspcc-dev/neofs-api-go/v2/rpc"
+	"github.com/nspcc-dev/neofs-api-go/v2/rpc/client"
+	v2session "github.com/nspcc-dev/neofs-api-go/v2/session"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	"github.com/nspcc-dev/neofs-sdk-go/object"
+	oid "github.com/nspcc-dev/neofs-sdk-go/object/id"
+	"github.com/nspcc-dev/neofs-sdk-go/session"
+	"github.com/nspcc-dev/neofs-sdk-go/token"
+)
+
+// PrmObjectSearch groups parameters of ObjectSearch operation.
+type PrmObjectSearch struct {
+	local bool
+
+	sessionSet bool
+	session    session.Token
+
+	bearerSet bool
+	bearer    token.BearerToken
+
+	cnrSet bool
+	cnr    cid.ID
+
+	filters object.SearchFilters
+}
+
+// MarkLocal tells the server to execute the operation locally.
+func (x *PrmObjectSearch) MarkLocal() {
+	x.local = true
+}
+
+// WithinSession specifies session within which the search query must be executed.
+//
+// Creator of the session acquires the authorship of the request.
+// This may affect the execution of an operation (e.g. access control).
+//
+// Must be signed.
+func (x *PrmObjectSearch) WithinSession(t session.Token) {
+	x.session = t
+	x.sessionSet = true
+}
+
+// WithBearerToken attaches bearer token to be used for the operation.
+//
+// If set, underlying eACL rules will be used in access control.
+//
+// Must be signed.
+func (x *PrmObjectSearch) WithBearerToken(t token.BearerToken) {
+	x.bearer = t
+	x.bearerSet = true
+}
+
+// InContainer specifies the container in which to look for objects.
+// Required parameter.
+func (x *PrmObjectSearch) InContainer(id cid.ID) {
+	x.cnr = id
+	x.cnrSet = true
+}
+
+// SetFilters sets filters by which to select objects. All container objects
+// match unset/empty filters.
+func (x *PrmObjectSearch) SetFilters(filters object.SearchFilters) {
+	x.filters = filters
+}
+
+// ResObjectSearch groups the final result values of ObjectSearch operation.
+type ResObjectSearch struct {
+	statusRes
+}
+
+// ObjectListReader is designed to read list of object identifiers from NeoFS system.
+//
+// Must be initialized using Client.ObjectSearch, any other usage is unsafe.
+type ObjectListReader struct {
+	cancelCtxStream context.CancelFunc
+
+	ctxCall contextCall
+
+	reqWritten bool
+
+	// initially bound to contextCall
+	bodyResp v2object.SearchResponseBody
+
+	tail []*v2refs.ObjectID
+}
+
+// UseKey specifies private key to sign the requests.
+// If key is not provided, then Client default key is used.
+func (x *ObjectListReader) UseKey(key ecdsa.PrivateKey) {
+	x.ctxCall.key = key
+}
+
+// Read reads another list of the object identifiers. Works similar to
+// io.Reader.Read but copies oid.ID and returns success flag instead of error.
+//
+// Failure reason can be received via Close.
+//
+// Panics if buf has zero length.
```

Comment:
suggest returning `0, false` to indicate that nothing happened to be more similar to `io.Reader`

Reply:
Everything is similar to `io.Reader` here, but not the same.